### ISSUE TITLE
Unexpected type widening during inference (behaves differently in Scala 2)

### ISSUE THREAD

**Issue by kitlangton** [lampepfl/dotty#14930]
## Compiler version

`3.1.1`

## Minimized code

Here's a [Scastie with the unexpected behavior in Scala 3](https://scastie.scala-lang.org/vH018w2TTBut0dqdNpDFIg). And here's a [Scastie with the expected behavior in Scala 2](https://scastie.scala-lang.org/m5PieJzmSiSjLtVevwVBfA).

Here's the gist:

```Scala
// Given a constructor that derives types from a function
def fromFunction[A, B, C](f: (A, B) => C)(implicit tagA: Tag[A], tagB: Tag[B]): Box[A with B] =
  Box(Map("a" -> tagA, "b" -> tagB))
 
// It behaves differently depending on whether or not the call site is given a type ascription
val box1 = fromFunction((int: Int, string: String) => true)
println(box2) // Infers A and B correctly

val box2: Box[Int with String] = fromFunction((int: Int, string: String) => true)
println(box3) // Infers A and B to both be (A & B)
```

## Output

```scala
 // In the case without the type ascription
Box(Map(a -> Tag[Int], b -> Tag[String]))

// In the case with the type ascription
Box(Map(a -> Tag[{String & Int}], b -> Tag[{String & Int}])) 
```

## Expectation

I would expect both cases to infer with a bias on the input type, which is clearly a function of type `(Int, String) => Boolean`. So I would expect both cases to print `Box(Map(a -> Tag[Int], b -> Tag[String]))`, as it does in the linked Scala 2 version.

_Concretely, this is used in `ZIO` to [generate "Layers" from functions](https://github.com/kitlangton/zio/blob/c4a3ff07f1cfee16ae8e5afbc3d613874a9cdd4b/core/shared/src/main/scala/zio/ZLayer.scala#L569)._

**Comment by smarter** [comment]
Minimized to not depend on external code:
```scala
class Tag[A]
class Box[-A]

class A {
  implicit val tagInt: Tag[Int] = ???
  implicit val tagString: Tag[String] = ???

  def fromFunction[A, B, C](f: (A, B) => C)(implicit tagA: Tag[A], tagB: Tag[B]): Box[A with B] = ???
  
  val box2 =
    fromFunction((int: Int, string: String) => true) // OK: Infers ?A := Int, ?B = String
  val box3: Box[Int with String] =
    fromFunction((int: Int, string: String) => true) // error: Infers ?A := Int & String, ?B = Int & String
}

```
When we get to the implicit parameter block we have as constraints:
```scala
     ?A >: Int & String <: Int
     ?B >: Int & String <: String
```
(the lower bound constraints come from the expected type, the upper bound constraints come from the lambda parameter types).

The logic used to instantiate type variables before implicit search is:
https://github.com/lampepfl/dotty/blob/f4822ac22bd7bc8eecb134ca5dc39585be9ff357/compiler/src/dotty/tools/dotc/typer/Inferencing.scala#L56-L65
Here, 1. applies: ?A is constrained from below to `Int & String` so we instantiate it to `Int & String`, same for ?B. The heuristic is that doing an implicit search with more precisely constrained types is more likely to yield the result you're looking for. It turns out that in this particular case this isn't true, but it's not clear to me if there's anything we can do to handle this case differently without breaking other things.

In conclusion, I don't recommend relying on type inference implementation details, especially when intersection/union types are involved :sweat_smile:

**Comment by smarter** [comment]
For reference, this isn't the first time that we have inference issues involving intersections in the expected type, I was able to significantly improve the situation in https://github.com/lampepfl/dotty/pull/8635, the key difference here is the interaction between type inference and implicit search which is poorly understood.
